core/src/services/webhdfs/writer.rs

```diff
@@ -41,14 +43,67 @@ impl WebhdfsWriter {
 }
 
 #[async_trait]
-impl oio::OneShotWrite for WebhdfsWriter {
+impl oio::MultipartUploadWrite for WebhdfsWriter {
     /// Using `bytes` instead of `vectored_bytes` to allow request to be redirected.
-    async fn write_once(&self, bs: &dyn WriteBuf) -> Result<()> {
-        let bs = bs.bytes(bs.remaining());
+    async fn write_once(&self, size: u64, body: AsyncBody) -> Result<()> {
+        let bs = match body {
+            AsyncBody::Empty => Bytes::from_static(&[]),
+            AsyncBody::Bytes(bs) => bs,
+            AsyncBody::ChunkedBytes(cb) => cb.bytes(cb.remaining()),
+            AsyncBody::Stream(stream) => {
+                // get the bytes from stream
+                stream.collect().await?
+            }
+        };
 
         let req = self.backend.webhdfs_create_object_request(
             &self.path,
-            Some(bs.len()),
+            Some(size as usize),
+            &self.op,
+            AsyncBody::Bytes(bs),
+        )?;
+
+        let resp = self.backend.client.send(req).await?;
+
+        let status = resp.status();
+        match status {
+            StatusCode::CREATED | StatusCode::OK => {
+                resp.into_body().consume().await?;
+                Ok(())
+            }
+            _ => Err(parse_error(resp).await?),
+        }
+    }
+
+    async fn initiate_part(&self) -> Result<String> {
+        Ok("".to_string())
+    }
+
+    /// Using `bytes` instead of `vectored_bytes` to allow request to be redirected.
+    async fn write_part(
+        &self,
+        _upload_id: &str,
+        part_number: usize,
+        size: u64,
+        body: AsyncBody,
+    ) -> Result<oio::MultipartUploadPart> {
+        let p = if part_number == 0 {
+            self.path.clone()
+        } else {
+            // Using MD5 to prevent overwriting existing files when writing.
+            format_md5(format!("{}-{}", self.path, part_number).as_bytes())
```

Comment:
We could write files to a temporary directory like `.opendal_tmp/` to hide partially written files from users. A new configuration option, such as `tmp_write_dir`, can be implemented for this purpose. The `write_can_multi` feature would then be available only if the user has enabled this setting.

Reply:
Webhdfs CONCAT files must in the same directory, tmp dir may not work